### ISSUE TITLE
Does this run?

### ISSUE THREAD

**Issue by ghallyb** [agrif/uvk5-rust#1]
Hi there. I was wondering if you actually got this running on a UV-K5? I'm interested in the project as a basis for some Rust development on the radio, but I was a bit apprehensive to try flashing it onto my radio given the note at the bottom of the Readme.

**Comment by agrif** [comment]
Yes! It runs on my UV-K5. The warning is there because all of this is patched together from the C firmware source and a machine-translated datasheet, and there are definitely ways to accidentally poke the device that are unrecoverable. However, after having worked with it for a while, in my estimation the radio is mostly unbrickable (unless you *really* try). Luckily, they are also cheap.

The main concern is overwriting the built-in bootloader, which most flashing tools (including mine) take some pains to prevent. The next concern is driving the radio in some way that damages the amplifier, but I haven't written any code to talk to the transmitter components. Still, the risk is there. I just want you to make an informed decision.

If you want to use this code, that's great! I've got a PAC that covers most of the low-level peripherals the UV-K5 uses, with some work on a HAL on top. It's good enough to do UART, talk to the keypad, the eeprom, and the lcd screen. I believe I was starting to work on the FM radio when I ran out of steam.

Documentation is *very* light but [a generated copy is available here](https://agrif.github.io/uvk5-rust/). Of particular note are the [dp32g030-hal](https://agrif.github.io/uvk5-rust/k5firmware/dp32g030_hal/index.html) and [k5board](https://agrif.github.io/uvk5-rust/k5firmware/k5board/index.html) crates. There are examples inside the `dp32g030-hal` and `k5board` crates that might help you get started. In general, the DP32G030 is an ARM Cortex-M0, similar to (but in no way identical to) the STM32 M0's. There's a lot of resources for embedded rust on STM32s, and those resources can be helpful. Just, we need to use our own PAC and HAL. My HAL is a poor imitation of the STM32F4 HAL crate.

If you have any questions, I'd be happy to try and answer them, time willing. I started this project to try to build cool things with a cool radio, so I'd be curious to see anything you build (even if it's like, Hello World on the LCD).

**Comment by agrif** [comment]
Oh and, two warnings that aren't relevant unless you start seriously poking around, but that are worth keeping in the back of your mind:

One, the stock firmware will stop functioning if the data in the EEPROM isn't what it expects. So, if you write to the EEPROM, either make sure to write to a location the stock firmware doesn't use, or take a full backup of what's on your radio first. I believe my flashing tool can do this (when the radio is loaded with stock or modified firmware).

Two, I have some half-baked code to erase and write to the flash memory on the DP32G030. This works, but there are *real* dragons there, and it would be quite easy to accidentally brick a radio by erasing the bootloader. None of this code is used in anything by default, so it'll only become an issue if you want to write to flash yourself. Be very careful writing to flash on the device itself.

**Comment by ghallyb** [comment]
Thanks for getting back to me and in such detail. I appreciate it. 

I purchased the radio with a view to using it to listen to a few channels in preparation for getting my HAM license. However, I realized that it scratches another itch in that it has a reasonable MCU, a screen, a keyboard and a battery in a cheap and sturdy package. To that end, I've ordered another one with the intention of hacking about on it and writing a couple of small, offline (as in making no use of the radio as of yet) applications. 

I am happy to share back anything I achieve, but I'm reasonably new to rust, so who knows what that will end up being or how long it will take.

I do have a small number of questions that hopefully you might be able to help with. Obviously, no rush.

I've had a good poke around the [Egzumer Firmware](https://github.com/egzumer/uv-k5-firmware-custom) to get my bearings. I was left with a few questions in relation to the bootloader (which you touched on in your last post). Mainly:

- **Are the egzumer and other popular firmwares generally replacing the bootloader or leaving the stock one in place?** 
  If I understand correctly from your last message, they are leaving the stock one in place (with the exception of a few projects that are specifically aimed at writing custom bootloaders).
- **Is "Flashing Mode" (ie, turning the radio on with PTT pressed and the handling of communication over the programming jack) implemented in the stock bootloader?** 
  Again, if I am understanding correctly, this functionality is part of the bootloader, so your (and others) implementations don't generally have to worry about this on the radio side? The firmware just needs to packaged correctly and put in the correct place?
- **Does the code in K5Tool directory largely do the same as the similarly named [K5TOOL found here](https://github.com/qrp73/K5TOOL)? That is to say, it packs, obfuscates and transfers the generated binaries.**

Thanks in advance for any help you are able to provide.

**Comment by agrif** [comment]
>Are the egzumer and other popular firmwares generally replacing the bootloader or leaving the stock one in place?

They leave the stock bootloader in place. It's a pretty good bootloader, and pretty well understood, and trying to overwrite it is one of the only reliable ways to brick the MCU. The bootloader uses a known good clock configuration and pauses if PTT is pressed, which is fantastic because a *bad* clock configuration with no pause is unrecoverable. Unlike STM32s, the DP32G030 can't be reset directly into debug, so any error that makes the MCU non-functional too fast can't be stopped.

>Is "Flashing Mode" (ie, turning the radio on with PTT pressed and the handling of communication over the programming jack) implemented in the stock bootloader?

Yes!

> Does the code in K5Tool directory largely do the same as the similarly named [K5TOOL found here](https://github.com/qrp73/K5TOOL)? That is to say, it packs, obfuscates and transfers the generated binaries.

Yes, and many many other things. Most notably, it understands the ELF binaries that rust produces and can flash them directly, and can attach to the serial port afterwards to show any output. It also has a few other useful bits, for example, it can simulate being a radio with a unix FIFO to talk with *other* flashing tools and see what bytes they send.

It uses the `k5lib` crate to speak the bootloader (and stock firmware's) protocol. The same crate can also be used on the device, from the other direction. I fear `k5lib` might be overengineered but I forget the details at the moment.
